internal/ghworkflow/ci.go

```diff
@@ -0,0 +1,182 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ghworkflow
+
+import (
+	"errors"
+	"io/ioutil"
+	"strings"
+
+	"golang.org/x/mod/modfile"
+)
+
+func ciWorkflow(cfg *Configuration) error {
+	goVersion := cfg.CI.GoVersion
+	if goVersion == "" {
+		filename := "go.mod"
+		data, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return err
+		}
+		f, err := modfile.Parse(filename, data, nil)
+		if err != nil {
+			return err
+		}
+		goVersion = f.Go.Version
+		if goVersion == "" {
+			return errors.New("could not find Go version from go.mod file, consider defining manually by setting githubWorkflows.ci.goVersion")
+		}
+	}
+	ignorePaths := cfg.Global.IgnorePaths
+	if cfg.CI.IgnorePaths != nil {
+		ignorePaths = cfg.CI.IgnorePaths
+	}
+
+	w := &workflow{
+		Name: "CI",
+		On:   eventTriggers(cfg.Global.DefaultBranch, ignorePaths),
+		Jobs: make(map[string]job),
+	}
+
+	makeOpts := ""
+	if cfg.Vendoring {
+		makeOpts = "GO_BUILDFLAGS='-mod vendor'"
+	}
+
+	// 01. Lint codebase.
+	lintJob := baseJobWithGo("Lint", goVersion)
+	lintJob.If = `!contains(github.event.head_commit.message, '[ci skip]')`
+	if cfg.GolangciLint {
+		// Don't need actions/cache here; golangci-lint has built-in caching.
+		lintJob.addStep(jobStep{
+			Name: "Run golangci-lint",
+			Uses: "golangci/golangci-lint-action@v2",
+			With: map[string]interface{}{
+				"version":              "latest",
+				"skip-go-installation": true,
+			},
+		})
+	} else {
+		if !cfg.Vendoring {
+			lintJob.addStep(cacheGoModules(true, lintJob.RunsOn))
+		}
+		lintJob.addStep(jobStep{
+			Name: "Run gofmt, golint, go vet",
+			Run:  stringsJoinAndTrimSpace([]string{makeOpts, "make", "static-check"}),
+		})
+	}
+	w.Jobs["lint"] = lintJob
+
+	buildTestOpts := buildTestJobOpts{
+		goVersion:    goVersion,
+		vendoring:    cfg.Vendoring,
+		runnerOSList: cfg.CI.RunnerOSList,
+	}
+
+	// 02. Make build.
+	buildTestOpts.name = "Build"
+	buildJob := buildOrTestBaseJob(buildTestOpts)
+	buildJob.Needs = []string{"lint"} // this is the <job_id> for the lint job
+	buildJob.addStep(jobStep{
+		Name: "Make build",
+		Run:  stringsJoinAndTrimSpace([]string{makeOpts, "make", "build-all"}),
+	})
+	w.Jobs["build"] = buildJob
+
+	// 03. Run tests and generate/upload test coverage.
+	buildTestOpts.name = "Test"
+	testJob := buildOrTestBaseJob(buildTestOpts)
+	testJob.Needs = []string{"build"} // this is the <job_id> for the build job
+	if cfg.CI.Postgres.Enabled {
+		version := defaultPostgresVersion
+		if cfg.CI.Postgres.Version != "" {
+			version = cfg.CI.Postgres.Version
+		}
+		testJob.Services = map[string]jobService{"postgres": {
+			Image: "postgres:" + version,
+			Env:   map[string]string{"POSTGRES_PASSWORD": "postgres"},
+			Ports: []string{"5432:5432"},
```

Comment:
The unit tests in all our Postgres-based services expect Postgres to be running on port 54321, which is where `testing/with-postgres-db.sh` also runs the Postgres on (in order to not conflict with a system-wide Postgres installation). We should probably use this port here as well, to avoid extra detection code in all the test harnesses.